src/connector/src/sink/postgres.rs

```diff
@@ -0,0 +1,625 @@
+    pub database: String,
+    pub table: String,
+    #[serde(default = "default_max_batch_rows")]
+    #[serde_as(as = "DisplayFromStr")]
+    pub max_batch_rows: usize,
+    pub r#type: String, // accept "append-only" or "upsert"
+}
+
+fn default_max_batch_rows() -> usize {
+    1024
+}
+
+impl PostgresConfig {
+    pub fn from_btreemap(properties: BTreeMap<String, String>) -> Result<Self> {
+        let config =
+            serde_json::from_value::<PostgresConfig>(serde_json::to_value(properties).unwrap())
+                .map_err(|e| SinkError::Config(anyhow!(e)))?;
+        if config.r#type != SINK_TYPE_APPEND_ONLY && config.r#type != SINK_TYPE_UPSERT {
+            return Err(SinkError::Config(anyhow!(
+                "`{}` must be {}, or {}",
+                SINK_TYPE_OPTION,
+                SINK_TYPE_APPEND_ONLY,
+                SINK_TYPE_UPSERT
+            )));
+        }
+        Ok(config)
+    }
+}
+
+#[derive(Debug)]
+pub struct PostgresSink {
+    pub config: PostgresConfig,
+    schema: Schema,
+    pk_indices: Vec<usize>,
+    is_append_only: bool,
+}
+
+impl PostgresSink {
+    pub fn new(
+        config: PostgresConfig,
+        schema: Schema,
+        pk_indices: Vec<usize>,
+        is_append_only: bool,
+    ) -> Result<Self> {
+        Ok(Self {
+            config,
+            schema,
+            pk_indices,
+            is_append_only,
+        })
+    }
+}
+
+impl TryFrom<SinkParam> for PostgresSink {
+    type Error = SinkError;
+
+    fn try_from(param: SinkParam) -> std::result::Result<Self, Self::Error> {
+        let schema = param.schema();
+        let config = PostgresConfig::from_btreemap(param.properties)?;
+        PostgresSink::new(
+            config,
+            schema,
+            param.downstream_pk,
+            param.sink_type.is_append_only(),
+        )
+    }
+}
+
+impl Sink for PostgresSink {
+    type Coordinator = DummySinkCommitCoordinator;
+    type LogSinker = LogSinkerOf<PostgresSinkWriter>;
+
+    const SINK_NAME: &'static str = POSTGRES_SINK;
+
+    async fn validate(&self) -> Result<()> {
+        if !self.is_append_only && self.pk_indices.is_empty() {
+            return Err(SinkError::Config(anyhow!(
+                "Primary key not defined for upsert Postgres sink (please define in `primary_key` field)")));
+        }
+
+        for field in self.schema.fields() {
+            check_data_type_compatibility(field.data_type())?;
+        }
+
+        // Verify pg table schema matches rw table schema, and pk indices are valid
+        let table_name = &self.config.table;
+        let connection_string = format!(
+            "host={} port={} user={} password={} dbname={}",
+            self.config.host,
+            self.config.port,
+            self.config.user,
+            self.config.password,
+            self.config.database
+        );
+        let (client, connection) =
+            tokio_postgres::connect(&connection_string, tokio_postgres::NoTls)
```

Comment:
May use a `sslmode` option to allow user config the ssl mode, can refer to `PostgresExternalTableReader`

Reply:
Looks like `tokio_postgres` does not support the full range of ssl_modes. https://github.com/sfackler/rust-postgres/issues/768.

I guess we have to use `sqlx::postgres` instead.

Reply:
Nevermind, fixed. https://github.com/risingwavelabs/risingwave/pull/19328/commits/e23758843d56832767b28218069049039b7eabe7